aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -521,12 +524,11 @@ Tensor trace_cpu(const Tensor& self) {
       sum += t_data[i * (t_stride_0 + t_stride_1)];
     }
 
-    // all integer types get promoted to kLong
-    if (result.scalar_type() == at::kLong) {
-      *result.data_ptr<int64_t>() = sum;
-    } else {
-      *result.data_ptr<scalar_t>() = sum;
-    }
+    c10::guts::if_constexpr<std::is_integral<accscalar_t>::value>(
+      // all integer types get promoted to kLong
+      [&] (auto _) { *result.data_ptr<int64_t>() = sum; },  // then-case, invalid for non-integral types
+      [&] (auto _) { *result.data_ptr<scalar_t>() = sum; }  // else-case, invalid for integral types
```

Comment:
cc. @ezyang the build still fails

Reply:
The compiler is eagerly type-checking the else-case even if the condition is true (and the other way round). You need to prevent the compiler from doing so, so that it can only type-check the branch that's actually called. You can do this by wrapping things into the `_` you get passed as an argument. See https://github.com/pytorch/pytorch/blob/master/c10/util/C%2B%2B17.h#L257